File: rs/day9/src/main.rs
use std::env;
use std::error;
use std::fs;

fn run(prog: &intcode::Program, input: i64) -> Result<i64, String> {
    let mut machine = intcode::Machine::new(prog);
    let mut output = Vec::new();
    loop {
        let state = machine.run()?;
        match state {
            intcode::State::Running => (),
            intcode::State::Reading => machine.write(input),
            intcode::State::Writing => output.push(machine.read()),
            intcode::State::Halted => break,
        }
    }
    if output.len() != 1 {
        Err(format!("bad output: {:?}", output))
    } else {
        Ok(output[0])
    }
}

fn main() -> Result<(), Box<dyn error::Error>> {
    let path = env::args().nth(1).ok_or("missing input path")?;
    let text = fs::read_to_string(&path)?;
    let prog = intcode::Program::new(&text)?;
    println!("{}", run(&prog, 1)?);
    println!("{}", run(&prog, 2)?);
    Ok(())
}

File: js/day9/main.js
"use strict";

const fs = require("fs");
const intcode = require("../intcode");

function run(prog, input) {
    const vm = new intcode.VM(
        prog,
        () => input,
        (x) => console.log(x)
    );
    vm.run();
}

function main(args) {
    const path = args[0];
    const file = fs.readFileSync(path, "ascii");
    const toks = file.split(",");
    const prog = toks.map((s) => parseInt(s, 10));
    run(prog, 1);
    run(prog, 2);
}

main(process.argv.slice(2));
